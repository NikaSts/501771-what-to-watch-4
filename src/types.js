import {string, number, shape, arrayOf, bool} from 'prop-types';

export const movieType = shape({
  id: number.isRequired,
  title: string.isRequired,
  runTime: number.isRequired,
  genre: string.isRequired,
  releaseDate: number.isRequired,
  poster: string.isRequired,
  previewImage: string.isRequired,
  backgroundImage: string.isRequired,
  backgroundColor: string.isRequired,
  ratingScore: number.isRequired,
  ratingLevel: string.isRequired,
  ratingCount: number.isRequired,
  paragraphs: arrayOf(string).isRequired,
  director: string.isRequired,
  stars: arrayOf(string).isRequired,
  preview: string.isRequired,
  isFavorite: bool.isRequired,
});

export const promoMovieType = shape({
  id: number.isRequired,
  title: string.isRequired,
  genre: string.isRequired,
  releaseDate: number.isRequired,
  poster: string.isRequired,
  backgroundImage: string.isRequired,
  backgroundColor: string.isRequired,
});

export const cardMovieType = shape({
  title: string.isRequired,
  image: string.isRequired,
});
