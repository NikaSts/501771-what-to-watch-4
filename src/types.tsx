export interface MovieType {
  id: number;
  title: string;
  runTime: number;
  genre: string;
  releaseDate: number;
  poster: string;
  previewImage: string;
  backgroundImage: string;
  backgroundColor: string;
  ratingScore: number;
  ratingLevel?: string;
  ratingCount: number;
  description: string;
  director: string;
  stars: Array<string>;
  preview: string;
  video: string;
  isFavorite: boolean;
}

export interface UserDataType {
  id: number;
  name: string;
  email: string;
  avatar: string;
}

export interface ReviewType {
  id: number;
  user: {
    id: number;
    name: string;
  };
  rating: number;
  comment: string;
  date: string;
}
