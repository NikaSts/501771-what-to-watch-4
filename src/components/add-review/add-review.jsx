import React from 'react';
import {Redirect, Link} from 'react-router-dom';
import {connect} from 'react-redux';
import {number, string, shape} from 'prop-types';

import {getMovie} from '../../store/movies/selectors';
import {getAuthorizationStatus, getUserData} from '../../store/user/selectors';
import {promoMovieType} from '../../types';
import {AppRoute, TEXTAREA_COLOR, AuthorizationStatus, Page} from '../../utils/consts';
import PageHeader from '../page-header/page-header';

const renderBreadCrumbs = (id, title) => (
  <nav className="breadcrumbs">
    <ul className="breadcrumbs__list">
      <li className="breadcrumbs__item">
        <Link to={`${AppRoute.MOVIE_PAGE}${id}`} className="breadcrumbs__link">
          {title}
        </Link>
      </li>
      <li className="breadcrumbs__item">
        <a className="breadcrumbs__link">Add review</a>
      </li>
    </ul>
  </nav>
);

const AddReview = ({authorizationStatus, id, movie, userData}) => {
  const isAuth = authorizationStatus === AuthorizationStatus.AUTH;

  if (!isAuth) {
    return <Redirect to={AppRoute.LOGIN} />;
  }
  const {backgroundColor, backgroundImage, title, poster} = movie;
  return (
    <section className="movie-card movie-card--full"
      style={{backgroundColor: `${backgroundColor}`}}>
      <div className="movie-card__header">
        <div className="movie-card__bg">
          <img src={backgroundImage} alt={title} />
        </div>

        <h1 className="visually-hidden">WTW</h1>

        <PageHeader
          currentPage={Page.REVIEW}
          isAuth={isAuth}
          userData={userData}>
          {renderBreadCrumbs(id, title)}
        </PageHeader>

        <div className="movie-card__poster movie-card__poster--small">
          <img src={poster} alt={`${title} poster`}
            widposterth="218" height="327" />
        </div>
      </div>

      <div className="add-review">
        <form action="#" className="add-review__form">
          <div className="rating">
            <div className="rating__stars">
              <input className="rating__input" id="star-1" type="radio" name="rating" value="1"/>
              <label className="rating__label" htmlFor="star-1">Rating 1</label>

              <input className="rating__input" id="star-2" type="radio" name="rating" value="2" />
              <label className="rating__label" htmlFor="star-2">Rating 2</label>

              <input className="rating__input" id="star-3" type="radio" name="rating" value="3"
                checked />
              <label className="rating__label" htmlFor="star-3">Rating 3</label>

              <input className="rating__input" id="star-4" type="radio" name="rating" value="4" />
              <label className="rating__label" htmlFor="star-4">Rating 4</label>

              <input className="rating__input" id="star-5" type="radio" name="rating" value="5" />
              <label className="rating__label" htmlFor="star-5">Rating 5</label>
            </div>
          </div>

          <div
            className="add-review__text"
            style={{backgroundColor: TEXTAREA_COLOR}}>
            <textarea className="add-review__textarea" name="review-text" id="review-text"
              placeholder="Review text"></textarea>
            <div className="add-review__submit">
              <button className="add-review__btn" type="submit">Post</button>
            </div>

          </div>
        </form>
      </div>
    </section>
  );
};

AddReview.propTypes = {
  authorizationStatus: string.isRequired,
  id: number,
  userData: shape({
    id: number.isRequired,
    name: string.isRequired,
    email: string.isRequired,
    avatar: string.isRequired,
  }),
  movie: promoMovieType,
};

const mapStateToProps = (state, props) => ({
  movie: getMovie(state, props.id),
  authorizationStatus: getAuthorizationStatus(state),
  userData: getUserData(state),
});

export {AddReview};
export default connect(mapStateToProps)(AddReview);
