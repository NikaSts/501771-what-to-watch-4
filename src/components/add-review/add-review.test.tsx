import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store';
import {Router} from 'react-router-dom';
import history from '../../history';
import {AddReview} from './add-review';
import {MovieType, UserDataType} from '../../types';


const mockStore = configureStore([]);

const movie: MovieType = {
  id: 1,
  title: `Movie`,
  runTime: 1,
  genre: ``,
  releaseDate: 1,
  poster: ``,
  previewImage: ``,
  backgroundImage: ``,
  backgroundColor: ``,
  ratingScore: 1,
  ratingLevel: ``,
  ratingCount: 1,
  description: ``,
  director: ``,
  stars: [``],
  preview: ``,
  video: ``,
  isFavorite: true,
};
const userData: UserDataType = {
  id: 1,
  name: ``,
  email: ``,
  avatar: ``,
};
const noop = () => {
  // Mock function for test props
};

it(`AddReview page should render with submit button active`, () => {
  const store = mockStore({
    MOVIES: {
      movie,
    },
    USER: {
      authorizationStatus: `AUTH`,
      user: userData,
    }
  });

  const tree = renderer
    .create(
        <Provider store={store}>
          <Router history={history}>
            <AddReview
              authorizationStatus={`AUTH`}
              id={1}
              user={userData}
              movie={movie}
              onFormSubmit={noop}
              onCommentChange={noop}
              onRatingChange={noop}
              isValid={true}
              isSending={false}
            />
          </Router>
        </Provider>, {
          createNodeMock: () => {
            return {};
          }
        })
      .toJSON();

  expect(tree).toMatchSnapshot();
});

it(`AddReview page should render with submit button disabled`, () => {
  const store = mockStore({
    MOVIES: {
      movie,
    },
    USER: {
      authorizationStatus: `AUTH`,
      user: userData,
    }
  });

  const tree = renderer
    .create(
        <Provider store={store}>
          <Router history={history}>
            <AddReview
              authorizationStatus={`AUTH`}
              id={1}
              user={userData}
              movie={movie}
              onFormSubmit={noop}
              onCommentChange={noop}
              onRatingChange={noop}
              isValid={false}
              isSending={false}
            />
          </Router>
        </Provider>, {
          createNodeMock: () => {
            return {};
          }
        })
      .toJSON();

  expect(tree).toMatchSnapshot();
});
