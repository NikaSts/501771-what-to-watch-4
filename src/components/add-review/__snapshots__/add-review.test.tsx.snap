// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddReview page should render with submit button active 1`] = `
<section
  className="movie-card movie-card--full"
  style={
    Object {
      "backgroundColor": "",
    }
  }
>
  <div
    className="movie-card__header"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="Movie"
        src=""
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header movie-card__head"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <nav
        className="breadcrumbs"
      >
        <ul
          className="breadcrumbs__list"
        >
          <li
            className="breadcrumbs__item"
          >
            <a
              className="breadcrumbs__link"
              href="/films/1"
              onClick={[Function]}
            >
              Movie
            </a>
          </li>
          <li
            className="breadcrumbs__item"
          >
            <a
              className="breadcrumbs__link"
            >
              Add review
            </a>
          </li>
        </ul>
      </nav>
      <div
        className="user-block"
      >
        <a
          href="/mylist"
          onClick={[Function]}
        >
          <div
            className="user-block__avatar"
          >
            <img
              alt="User avatar"
              height="63"
              src="https://4.react.pages.academy"
              width="63"
            />
          </div>
        </a>
      </div>
    </header>
    <div
      className="movie-card__poster movie-card__poster--small"
    >
      <img
        alt="Movie poster"
        height="327"
        src=""
        width="218"
      />
    </div>
  </div>
  <div
    className="add-review"
  >
    <form
      action="#"
      className="add-review__form"
      onSubmit={[Function]}
    >
      <div
        className="rating"
      >
        <div
          className="rating__stars"
          onChange={[Function]}
        >
          <input
            className="rating__input"
            disabled={false}
            id="star-1"
            name="rating"
            type="radio"
            value="1"
          />
          <label
            className="rating__label"
            htmlFor="star-1"
          >
            Rating 1
          </label>
          <input
            className="rating__input"
            disabled={false}
            id="star-2"
            name="rating"
            type="radio"
            value="2"
          />
          <label
            className="rating__label"
            htmlFor="star-2"
          >
            Rating 2
          </label>
          <input
            className="rating__input"
            defaultChecked={true}
            disabled={false}
            id="star-3"
            name="rating"
            type="radio"
            value="3"
          />
          <label
            className="rating__label"
            htmlFor="star-3"
          >
            Rating 3
          </label>
          <input
            className="rating__input"
            disabled={false}
            id="star-4"
            name="rating"
            type="radio"
            value="4"
          />
          <label
            className="rating__label"
            htmlFor="star-4"
          >
            Rating 4
          </label>
          <input
            className="rating__input"
            disabled={false}
            id="star-5"
            name="rating"
            type="radio"
            value="5"
          />
          <label
            className="rating__label"
            htmlFor="star-5"
          >
            Rating 5
          </label>
        </div>
      </div>
      <div
        className="add-review__text"
        style={
          Object {
            "backgroundColor": "rgba(255, 255, 255, 0.3)",
          }
        }
      >
        <textarea
          className="add-review__textarea"
          id="review-text"
          maxLength={400}
          minLength={50}
          name="review-text"
          onInput={[Function]}
          placeholder="Review text"
        />
        <div
          className="add-review__submit"
        >
          <button
            className="add-review__btn"
            disabled={false}
            type="submit"
          >
            Post
          </button>
        </div>
      </div>
    </form>
  </div>
</section>
`;

exports[`AddReview page should render with submit button disabled 1`] = `
<section
  className="movie-card movie-card--full"
  style={
    Object {
      "backgroundColor": "",
    }
  }
>
  <div
    className="movie-card__header"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="Movie"
        src=""
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header movie-card__head"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <nav
        className="breadcrumbs"
      >
        <ul
          className="breadcrumbs__list"
        >
          <li
            className="breadcrumbs__item"
          >
            <a
              className="breadcrumbs__link"
              href="/films/1"
              onClick={[Function]}
            >
              Movie
            </a>
          </li>
          <li
            className="breadcrumbs__item"
          >
            <a
              className="breadcrumbs__link"
            >
              Add review
            </a>
          </li>
        </ul>
      </nav>
      <div
        className="user-block"
      >
        <a
          href="/mylist"
          onClick={[Function]}
        >
          <div
            className="user-block__avatar"
          >
            <img
              alt="User avatar"
              height="63"
              src="https://4.react.pages.academy"
              width="63"
            />
          </div>
        </a>
      </div>
    </header>
    <div
      className="movie-card__poster movie-card__poster--small"
    >
      <img
        alt="Movie poster"
        height="327"
        src=""
        width="218"
      />
    </div>
  </div>
  <div
    className="add-review"
  >
    <form
      action="#"
      className="add-review__form"
      onSubmit={[Function]}
    >
      <div
        className="rating"
      >
        <div
          className="rating__stars"
          onChange={[Function]}
        >
          <input
            className="rating__input"
            disabled={false}
            id="star-1"
            name="rating"
            type="radio"
            value="1"
          />
          <label
            className="rating__label"
            htmlFor="star-1"
          >
            Rating 1
          </label>
          <input
            className="rating__input"
            disabled={false}
            id="star-2"
            name="rating"
            type="radio"
            value="2"
          />
          <label
            className="rating__label"
            htmlFor="star-2"
          >
            Rating 2
          </label>
          <input
            className="rating__input"
            defaultChecked={true}
            disabled={false}
            id="star-3"
            name="rating"
            type="radio"
            value="3"
          />
          <label
            className="rating__label"
            htmlFor="star-3"
          >
            Rating 3
          </label>
          <input
            className="rating__input"
            disabled={false}
            id="star-4"
            name="rating"
            type="radio"
            value="4"
          />
          <label
            className="rating__label"
            htmlFor="star-4"
          >
            Rating 4
          </label>
          <input
            className="rating__input"
            disabled={false}
            id="star-5"
            name="rating"
            type="radio"
            value="5"
          />
          <label
            className="rating__label"
            htmlFor="star-5"
          >
            Rating 5
          </label>
        </div>
      </div>
      <div
        className="add-review__text"
        style={
          Object {
            "backgroundColor": "rgba(255, 255, 255, 0.3)",
          }
        }
      >
        <textarea
          className="add-review__textarea"
          id="review-text"
          maxLength={400}
          minLength={50}
          name="review-text"
          onInput={[Function]}
          placeholder="Review text"
        />
        <div
          className="add-review__submit"
        >
          <button
            className="add-review__btn"
            disabled={true}
            type="submit"
          >
            Post
          </button>
        </div>
      </div>
    </form>
  </div>
</section>
`;
